File: LeetcodeProblems/blind-75/graph/pacific-atlantic-water-flow/solution.java
// Pacific Atlantic Water Flow - LeetCode #417

import java.util.*;

class Solution {
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        List<List<Integer>> result = new ArrayList<>();
        if (heights == null || heights.length == 0 || heights[0].length == 0) return result;
        
        int m = heights.length;
        int n = heights[0].length;
        
        boolean[][] pacific = new boolean[m][n];
        boolean[][] atlantic = new boolean[m][n];
        
        // DFS from borders
        for (int i = 0; i < m; i++) {
            dfs(heights, pacific, Integer.MIN_VALUE, i, 0);
            dfs(heights, atlantic, Integer.MIN_VALUE, i, n - 1);
        }
        
        for (int j = 0; j < n; j++) {
            dfs(heights, pacific, Integer.MIN_VALUE, 0, j);
            dfs(heights, atlantic, Integer.MIN_VALUE, m - 1, j);
        }
        
        // Find intersection
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (pacific[i][j] && atlantic[i][j]) {
                    result.add(Arrays.asList(i, j));
                }
            }
        }
        
        return result;
    }
    
    private void dfs(int[][] heights, boolean[][] visited, int height, int x, int y) {
        int m = heights.length;
        int n = heights[0].length;
        
        if (x < 0 || x >= m || y < 0 || y >= n || visited[x][y] || heights[x][y] < height) return;
        
        visited[x][y] = true;
        
        dfs(heights, visited, heights[x][y], x - 1, y);
        dfs(heights, visited, heights[x][y], x + 1, y);
        dfs(heights, visited, heights[x][y], x, y - 1);
        dfs(heights, visited, heights[x][y], x, y + 1);
    }
}

File: LeetcodeProblems/blind-75/graph/pacific-atlantic-water-flow/solution.py
# Pacific Atlantic Water Flow - LeetCode #417

from typing import List

class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        result = []
        if not heights or not heights[0]:
            return result
        
        m = len(heights)
        n = len(heights[0])
        
        pacific = [[False] * n for _ in range(m)]
        atlantic = [[False] * n for _ in range(m)]
        
        # DFS from borders
        for i in range(m):
            self.dfs(heights, pacific, float('-inf'), i, 0)
            self.dfs(heights, atlantic, float('-inf'), i, n - 1)
        
        for j in range(n):
            self.dfs(heights, pacific, float('-inf'), 0, j)
            self.dfs(heights, atlantic, float('-inf'), m - 1, j)
        
        # Find intersection
        for i in range(m):
            for j in range(n):
                if pacific[i][j] and atlantic[i][j]:
                    result.append([i, j])
        
        return result
    
    def dfs(self, heights: List[List[int]], visited: List[List[bool]], height: int, x: int, y: int):
        m = len(heights)
        n = len(heights[0])
        
        if x < 0 or x >= m or y < 0 or y >= n or visited[x][y] or heights[x][y] < height:
            return
        
        visited[x][y] = True
        
        self.dfs(heights, visited, heights[x][y], x - 1, y)
        self.dfs(heights, visited, heights[x][y], x + 1, y)
        self.dfs(heights, visited, heights[x][y], x, y - 1)
        self.dfs(heights, visited, heights[x][y], x, y + 1)
